### DISCUSSION TITLE
Scripting support?

### DISCUSSION THREAD

**Discussion by joao678** [samuelgr/Xidi#46] [Ideas]
Is it possible for a dll like this to load an external file and use a scripting language to interact with it? I'll give some examples:

Games like FlatOut 2 (a racing game) have very poor/basic controller support:

- You can set acceleration/brake/handling just fine by using any controller axis.
- However the menus can only be controlled by keyboard, so let's say I'm in a race and I set the B button as both handbrake and the menu back command (ESC key on keyboard). Everytime I press B while in a race the game whould pause, which can be frustrating, and you'll have to do some kind of workaround.
- Some controls are handled incorrectly, for example in the stunt mode instead of using the left stick being mapped to up/down/left/right movements like on PS2, the PC version uses acceleration/brake for both up/down and the handling axis for left/right, which can be confusing

Another example is Prince Of Persia: Sands of Time:

- This one is not too bad however there is one big issue
- The entire menu of the game can only be controlled using the mouse
- Since Xidi already supports moving the mouse and pressing mouse buttons maybe we can find a way to switch between two profiles if the game is paused or not

I suggest that maybe add some kind of scripting language to do some more advanced stuff, maybe something like Lua(maybe this one is easy to implement), Python or even JavaScript but of course keep it completely optional by still using the .ini files, I'll give you an example on how we can fix FlatOut 2 by doing some pseudo code:

```javascript
let isOnMenu = false,
    isDriving = false,
    isOnStuntMode = false;

while(true) {
    UpdateXidi();
    isOnMenu = ReadAddress("0x0000123"); //return an value from an address maybe a 1 or 0 in this case
    isOnStuntMode = ReadAddress("0x0000456");
    isDriving = ReadAddress("0x0000789");

    //"realController" represents the real xbox controller by player index and "xidiController" represents the virtual DirecInput controller created by xidi
    xidiController[0].btn1 = realController[0].A;
    /*
    ...
    not adding every button this is just a demo
    here comes the cool part
    */
    xidiController[0].btn2 = isOnMenu === true ? keyboard["ESC"] : realController[0].B;
    
    if(isOnStuntMode && !isDriving) {
        if (realController[0].LY > 0) xidiController[0].zRot = realController[0].LY;
        if (realController[0].LY < 0) xidiController[0].zAxis = realController[0].LY;
    }
}
```

**Comment by samuelgr** [comment]
While not currently supported in Xidi, it is possible to imagine a DLL loading and running a script like this. Also see issue #40 for another thread talking about making Xidi externally programmable via an API.

I actually really like the idea of making a Xidi API for plugin/patch developers to use to control things like which mapper is in use at any given time, among possible other things. I do not foresee myself taking the time to write a script interpreter inside Xidi, but someone else could certainly use the API to do so as an external plugin module.

**Reply by joao678** [reply]
Oh yeah, some kind of api whould be ideal in order to fully support all edge use cases, I don't know C++ but I whould be more than happy to contribute to anything if I could, xidi is a total game changer for Steam, as the xinput implementation they did is amazing and xidi plays really nice with multiple controllers on steam deck with the easy reordering that steam provides. I'm hoping some kind of api comes out, you have my full support!

**Comment by joao678** [comment]
Another contribution you might consider is a system wide implementation with device hiding, like what [devreorder](https://github.com/briankendall/devreorder) does (see the "Apply to your entire system" section) its a pretty intresting project and I think it fits well with xidi, maybe if combining both whould be the ultimate DirectInput solution to all of our problems.

Edit: forgot to mention btw, allowing for a systemwide solution like this can benefit from other mods in games to freely use the dinput8.dll from the game's folder (such as widescreen mods), also..is there any way to use two dinput8.dlls with the same game?
